querqy-core/src/main/java/querqy/rewrite/ContextAwareQueryRewriter.java

```diff
@@ -1,39 +1,118 @@
 /**
- * 
+ *
  */
 package querqy.rewrite;
 
+import querqy.infologging.InfoLoggingContext;
+import querqy.model.AbstractNodeVisitor;
 import querqy.model.ExpandedQuery;
+import querqy.model.Node;
+
+import java.util.HashSet;
+import java.util.IdentityHashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
 
 /**
  * If a {@link QueryRewriter} implements this interface, the {@link #rewrite(ExpandedQuery, SearchEngineRequestAdapter)}
  * method is called instead of {@link #rewrite(ExpandedQuery)} in the query rewriting chain. The Map argument can be
- * used to pass data between rewriters and to the consumer of the rewrite chain.
- * 
+ * used to pass data between rewriters and to the consumer of the rewrite chain. It also gives you the easy possibility to
+ * track the applied rules in a rewriter.
+ *
  * @author René Kriegler, @renekrie
  *
  */
-public interface ContextAwareQueryRewriter extends QueryRewriter {
+public abstract class ContextAwareQueryRewriter extends AbstractNodeVisitor<Node> implements QueryRewriter {
 
     /**
      * Name of the context key that acts as a flag to request debug information.
      */
-    String CONTEXT_KEY_DEBUG_ENABLED = "querqy.debug.rewrite.isdebug";
+    public static final String CONTEXT_KEY_DEBUG_ENABLED = "querqy.debug.rewrite.isdebug";
 
     /**
      * The key under which debug information should be collected if CONTEXT_KEY_DEBUG_ENABLED is set.
      */
-    String CONTEXT_KEY_DEBUG_DATA = "querqy.debug.rewrite.data";
+    public static final String CONTEXT_KEY_DEBUG_DATA = "querqy.debug.rewrite.data";
+
+    /**
+     * Marker in the infologging for all applied rules.
+     */
+    private static final String APPLIED_RULES = "APPLIED_RULES";
+
+    protected SearchEngineRequestAdapter searchEngineRequestAdapter;
```

Comment:
We should avoid making the SearchEngineRequest adapter a member variable. We have no guarantee that rewriter objects are created per request.